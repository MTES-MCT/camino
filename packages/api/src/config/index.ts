// virtualHost est d√©fini seulement si
// l'application tourne dans un container docker
// voir le fichier DockerFile

const debug = process.env.NODE_DEBUG === 'true'
const virtualHost = process.env.VIRTUAL_HOST
const port = Number(process.env.PORT)
const url = virtualHost ? `http://${virtualHost}/` : `http://localhost:${port}/`

export { port, url, debug }
