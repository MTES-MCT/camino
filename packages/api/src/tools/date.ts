import { CaminoDate, toCaminoDate } from 'camino-common/src/date.js'

export const dateValidate = (str: string | undefined | null) => {
  if (!str) return 'Date manquante'

  const date = new Date(str)

  if (date.toString() === 'Invalid Date') {
    return 'Date invalide'
  }

  return null
}

// ajoute une durée en jours à une string au format YYYY-MM-DD
export const dateAddDays = (date: CaminoDate, days: number): CaminoDate => {
  const [y, m, d] = date.split('-')

  return toCaminoDate(new Date(+y, +m - 1, +d + days))
}

// ajoute une durée en mois à une string au format YYYY-MM-DD
export const dateAddMonths = (date: CaminoDate, months: number): CaminoDate => {
  const [y, m, d] = date.split('-')

  return toCaminoDate(new Date(+y, +m - 1 + months, +d))
}

// calcul le nombre de mois entre 2 dates
export const datesSubtract = (dateDebut: string, dateFin: string) => {
  const [yDebut, mDebut] = dateDebut.split('-')
  const [yFin, mFin] = dateFin.split('-')

  return +yFin * 12 + +mFin - (+yDebut * 12 + +mDebut)
}
