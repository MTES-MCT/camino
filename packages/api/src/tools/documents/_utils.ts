import { Index } from '../../types.js'

const hashGet = (str: string) => str.split('-').pop()

export const matchFuzzy = (name: string, index: Index<any>, partGet = hashGet) => {
  const hash = name.split('-').pop()

  return Object.keys(index).reduce((r: string[], key) => {
    // on ne garde pas les matches entiers pendant un fuzzy
    if (key === name) {
      return r
    }

    const part = partGet(key)

    if (part === hash) {
      r.push(key)
    }

    return r
  }, [])
}
