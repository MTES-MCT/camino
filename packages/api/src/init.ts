import { knexInit } from './knex'
import { knexConfig } from './knex/config'

knexInit(knexConfig)
