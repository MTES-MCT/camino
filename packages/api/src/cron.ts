import './init'
import { job } from 'cron'
import daily from './business/daily'
import documentsCheck from './tools/documents/check'
import { matomoCacheInit } from './tools/api-matomo'
import demarchesDefinitionsCheck from './tools/demarches/definitions-check'
import { consoleOverride } from './config/logger'
import { mailjetSend } from './tools/api-mailjet/emails'
import { readFileSync, writeFileSync, createWriteStream } from 'fs'
import { titreTypeDemarcheTypeEtapeTypeCheck } from './tools/demarches/tde-check'
import { etapeStatutCheck } from './tools/demarches/etape-statut-check'
import { documentsClean } from './tools/documents/clean'
import * as Console from 'console'

const logFile = '/tmp/cron.log'

const output = createWriteStream(logFile)
const logger = new Console.Console({ stdout: output })
// eslint-disable-next-line no-console
console.log = logger.log
consoleOverride(false)

const tasks = async () => {
  console.info('Cron quotidien : démarrage')
  // Réinitialise les logs qui seront envoyés par email
  writeFileSync(logFile, '')

  await daily()
  await documentsClean()
  await documentsCheck()
  await demarchesDefinitionsCheck()
  await titreTypeDemarcheTypeEtapeTypeCheck()
  await etapeStatutCheck()

  try {
    await matomoCacheInit()
  } catch (e) {
    console.error('API Matomo inaccessible', e)
  }

  const emailBody = `Résultats de ${process.env.ENV} \n${readFileSync(
    logFile
  ).toString()}`
  await mailjetSend([process.env.ADMIN_EMAIL!], {
    Subject: `[Camino][${process.env.ENV}] Résultats du daily`,
    'Text-part': emailBody
  })
  console.info('Cron quotidien : terminé')
}

job(
  // cronTime
  '00 00 04 * * *',
  // onTick
  tasks,
  //  onComplete
  null,
  // start
  true,
  // timezone
  'Europe/Paris',
  // context
  null,
  // runOnInit
  false
  // utcOffset
  // unrefTimeout
)
