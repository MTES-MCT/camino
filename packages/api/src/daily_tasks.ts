import './init'
import daily from './business/daily'
import documentsCheck from './tools/documents/check'
import { matomoCacheInit } from './tools/api-matomo'
import demarchesDefinitionsCheck from './tools/demarches/definitions-check'
import { consoleOverride } from './config/logger'
import { mailjetSend } from './tools/api-mailjet/emails'
import { readFileSync, writeFileSync, createWriteStream } from 'fs'
import { titreTypeDemarcheTypeEtapeTypeCheck } from './tools/demarches/tde-check'
import { etapeStatutCheck } from './tools/demarches/etape-statut-check'
import { documentsClean } from './tools/documents/clean'
import * as Console from 'console'

const logFile = '/tmp/cron.log'

const output = createWriteStream(logFile)
const logger = new Console.Console({ stdout: output, stderr: output })
// eslint-disable-next-line no-console
console.log = logger.log
consoleOverride(false)

const tasks = async () => {
  console.info('Tâches quotidiennes : démarrage')
  // Réinitialise les logs qui seront envoyés par email
  writeFileSync(logFile, '')
  try {
    await daily()
    await documentsClean()
    await documentsCheck()
    await demarchesDefinitionsCheck()
    await titreTypeDemarcheTypeEtapeTypeCheck()
    await etapeStatutCheck()
  } catch (e) {
    console.error('Erreur durant le daily', e)
  }

  try {
    await matomoCacheInit()
  } catch (e) {
    console.error('API Matomo inaccessible', e)
  }

  const emailBody = `Résultats de ${process.env.ENV} \n${readFileSync(
    logFile
  ).toString()}`
  await mailjetSend([process.env.ADMIN_EMAIL!], {
    Subject: `[Camino][${process.env.ENV}] Résultats du daily`,
    'Text-part': emailBody
  })
  console.info('Tâches quotidiennes : terminé')
}

tasks()
  .then(() => {
    output.end()
    process.exit()
  })
  .catch(e => {
    console.error('Erreur', e)
    output.end()
    process.exit(1)
  })
