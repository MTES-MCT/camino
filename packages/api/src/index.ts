/**
 * Camino API, le cadastre minier num√©rique ouvert
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import './init.js'
import compression from 'compression'
import cors from 'cors'
import express from 'express'
import rateLimit from 'express-rate-limit'
import * as Sentry from '@sentry/node'

import { port, url } from './config/index.js'
import { restWithPool } from './server/rest.js'
import { graphql } from './server/graphql.js'
import { authJwt } from './server/auth-jwt.js'
import { authBasic } from './server/auth-basic.js'
import { restUpload, graphqlUpload, uploadAllowedMiddleware } from './server/upload.js'
import { databaseInit } from './database/init.js'

import { consoleOverride } from './config/logger.js'
import { filesInit } from './config/files.js'
import { geoSystemesInit } from './config/proj4.js'
import { userLoader } from './server/user-loader.js'
import { connectedCatcher } from './server/connected-catcher.js'
import cookieParser from 'cookie-parser'

consoleOverride()
geoSystemesInit()
filesInit().then(() => {
  databaseInit().then(() => {
    const app = express()
    app.disable('x-powered-by')

    if (process.env.API_SENTRY_URL) {
      Sentry.init({
        dsn: process.env.API_SENTRY_URL,
        environment: process.env.ENV === 'prod' ? 'production' : process.env.ENV,
      })
      app.use(Sentry.Handlers.requestHandler())
    }

    const limiter = rateLimit({
      windowMs: 60 * 1000, // 1 minute
      max: 100, // Limit each IP to 100 requests per `window` (here, per 1 minute)
      standardHeaders: true, // Return rate limit info in the `RateLimit-*` headers
      legacyHeaders: false, // Disable the `X-RateLimit-*` headers
      skip: (request: any, _response: any) => {
        // On n'applique pas de rate limiting sur le televersement des fichiers
        return request.url.startsWith('/televersement')
      },
    })

    app.use(cors({ credentials: true, exposedHeaders: ['Content-disposition'] }), compression(), limiter, authJwt, authBasic, userLoader, cookieParser(), connectedCatcher)

    app.get('/stream/version', async (_req, res) => {
      const headers = {
        'Content-Type': 'text/event-stream',
        Connection: 'keep-alive',
        'Cache-Control': 'no-cache',
      }

      res.writeHead(200, headers)
      res.write(`id: ${Date.now()}\n`)
      res.write(`event: version\n`)
      res.write(`data: ${process.env.APPLICATION_VERSION}\n\n`)
      res.flush()
    })
    app.use(express.urlencoded({ extended: true }), express.json(), restWithPool())

    app.use('/televersement', uploadAllowedMiddleware, restUpload())

    app.use('/', graphqlUpload, graphql)

    if (process.env.API_SENTRY_URL) {
      app.use(Sentry.Handlers.errorHandler())
    }

    app.listen(port, () => {
      console.info('')
      console.info('URL:', url)
      console.info('ENV:', process.env.ENV)
      console.info('NODE_ENV:', process.env.NODE_ENV)

      if (process.env.NODE_DEBUG === 'true') {
        console.warn('NODE_DEBUG:', process.env.NODE_DEBUG)
      }
      console.info('')
    })
  })
})
