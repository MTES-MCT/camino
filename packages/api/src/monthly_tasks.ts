import './init'
import { consoleOverride } from './config/logger.js'
import { mailjetSend } from './tools/api-mailjet/emails.js'
import { readFileSync, writeFileSync, createWriteStream } from 'fs'
import * as Console from 'console'
import { monthly } from './business/monthly.js'

const logFile = '/tmp/monthly.log'

const output = createWriteStream(logFile)
const logger = new Console.Console({ stdout: output, stderr: output })
// eslint-disable-next-line no-console
console.log = logger.log
consoleOverride(false)

const tasks = async () => {
  console.info('Tâches mensuelles : démarrage')
  // Réinitialise les logs qui seront envoyés par email
  writeFileSync(logFile, '')
  try {
    await monthly()
  } catch (e) {
    console.error('Erreur durant le monthly', e)
  }

  const emailBody = `Résultats de ${process.env.ENV} \n${readFileSync(
    logFile
  ).toString()}`
  await mailjetSend([process.env.ADMIN_EMAIL!], {
    Subject: `[Camino][${process.env.ENV}] Résultats du monthly`,
    'Text-part': emailBody
  })
  console.info('Tâches mensuelles : terminé')
}

tasks()
  .then(() => {
    output.end()
    process.exit()
  })
  .catch(e => {
    console.error('Erreur', e)
    output.end()
    process.exit(1)
  })
