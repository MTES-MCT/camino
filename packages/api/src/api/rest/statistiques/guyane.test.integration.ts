import { restCall } from '../../../../tests/_utils/index'
import { dbManager } from '../../../../tests/db-manager'
import { expect, test, afterAll, beforeAll, vi } from 'vitest'
import type { Pool } from 'pg'
import { caminoAnneeValidator } from 'camino-common/src/date'

console.info = vi.fn()
console.error = vi.fn()
let dbPool: Pool

beforeAll(async () => {
  const { pool } = await dbManager.populateDb()
  dbPool = pool
})

afterAll(async () => {
  await dbManager.closeKnex()
})

test('peut récupérer les statistiques de Guyane', async () => {
  const tested = await restCall(dbPool, '/rest/statistiques/guyane/:annee', { annee: caminoAnneeValidator.parse('2023') }, undefined)

  expect(tested.statusCode).toBe(200)
  expect(tested.body).toMatchSnapshot()
})
