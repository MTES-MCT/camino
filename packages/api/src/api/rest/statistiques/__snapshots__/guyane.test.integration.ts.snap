// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`peut récupérer les statistiques de Guyane 1`] = `
{
  "annees": [
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2018",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 1320,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2019",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 0,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2020",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 0,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2021",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 0,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2022",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 0,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
    {
      "activitesDeposesQuantite": 0,
      "activitesDeposesRatio": 0,
      "annee": "2023",
      "carburantConventionnel": 0,
      "carburantDetaxe": 0,
      "effectifs": 0,
      "environnementCout": 0,
      "mercure": 0,
      "orNet": 0,
      "titresArm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresAxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresCxm": {
        "quantite": 0,
        "surface": 0,
      },
      "titresPrm": {
        "quantite": 0,
        "surface": 0,
      },
    },
  ],
  "arm": {
    "depot": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "octroiEtProlongation": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "refusees": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "surface": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
  },
  "axm": {
    "depot": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "octroiEtProlongation": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "refusees": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "surface": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
  },
  "cxm": {
    "depot": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "octroiEtProlongation": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "refusees": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "surface": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
  },
  "prm": {
    "depot": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "octroiEtProlongation": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "refusees": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
    "surface": {
      "2017": 0,
      "2018": 0,
      "2019": 0,
      "2020": 0,
      "2021": 0,
      "2022": 0,
      "2023": 0,
    },
  },
  "surfaceExploitation": 0,
  "surfaceExploration": 0,
  "titresArm": 0,
  "titresAxm": 0,
  "titresCxm": 0,
  "titresPrm": 0,
}
`;
